pkg/parser/whitelist.go

```diff
@@ -19,91 +19,103 @@ type Whitelist struct {
 	B_Cidrs []*net.IPNet
 	Exprs   []string `yaml:"expression,omitempty"`
 	B_Exprs []*ExprWhitelist
-	Node    *Node `yaml:"-"` // Pointer to the node containing this whitelist
 }
 
-func (W Whitelist) ContainsIPLists() bool {
-	return len(W.B_Ips) > 0 || len(W.B_Cidrs) > 0
+type ExprWhitelist struct {
+	Filter       *vm.Program
+	ExprDebugger *exprhelpers.ExprDebugger // used to debug expression by printing the content of each variable of the expression
 }
 
-func (W Whitelist) ContainsExprLists() bool {
-	return len(W.B_Exprs) > 0
+func (n *Node) ContainsIPLists() bool {
+	return len(n.Whitelist.B_Ips) > 0 || len(n.Whitelist.B_Cidrs) > 0
 }
 
-func (W Whitelist) Check(srcs []net.IP, cachedExprEnv map[string]interface{}) (bool, bool, error) {
+func (n *Node) ContainsExprLists() bool {
+	return len(n.Whitelist.B_Exprs) > 0
+}
+
+func (n *Node) CheckIPsWL(srcs []net.IP) (bool, bool) {
 	isWhitelisted := false
-	hasWhitelist := false
-	var err error
-	var output interface{}
-	if W.ContainsIPLists() {
-		for _, src := range srcs {
-			if isWhitelisted {
+	if !n.ContainsIPLists() {
+		return isWhitelisted, false
+	}
+	for _, src := range srcs {
+		if isWhitelisted {
+			break
+		}
+		for _, v := range n.Whitelist.B_Ips {
+			if v.Equal(src) {
+				n.Logger.Debugf("Event from [%s] is whitelisted by IP (%s), reason [%s]", src, v, n.Whitelist.Reason)
+				isWhitelisted = true
 				break
 			}
-			for _, v := range W.B_Ips {
-				if v.Equal(src) {
-					W.Node.Logger.Debugf("Event from [%s] is whitelisted by IP (%s), reason [%s]", src, v, W.Reason)
-					isWhitelisted = true
-					break
-				}
-				W.Node.Logger.Tracef("whitelist: %s is not eq [%s]", src, v)
-			}
-			for _, v := range W.B_Cidrs {
-				if v.Contains(src) {
-					W.Node.Logger.Debugf("Event from [%s] is whitelisted by CIDR (%s), reason [%s]", src, v, W.Reason)
-					isWhitelisted = true
-					break
-				}
-				W.Node.Logger.Tracef("whitelist: %s not in [%s]", src, v)
+			n.Logger.Tracef("whitelist: %s is not eq [%s]", src, v)
+		}
+		for _, v := range n.Whitelist.B_Cidrs {
+			if v.Contains(src) {
+				n.Logger.Debugf("Event from [%s] is whitelisted by CIDR (%s), reason [%s]", src, v, n.Whitelist.Reason)
+				isWhitelisted = true
+				break
 			}
+			n.Logger.Tracef("whitelist: %s not in [%s]", src, v)
 		}
-		hasWhitelist = true
+	}
+	return isWhitelisted, true
+}
+
+func (n *Node) CheckExprWL(cachedExprEnv map[string]interface{}) (bool, bool, error) {
+	err := error(nil)
+	output := interface{}(nil)
+
+	isWhitelisted := false
+
+	if !n.ContainsExprLists() {
+		return false, false, nil
```

Comment:
There is a potential bug when an IP doesnt pass the IP based whitelist but doesnt have expressions it will make hasWhitelist to false.